src/AzureClient/AzureClient.cs

```diff
@@ -77,98 +69,33 @@ private string ValidExecutionTargetsDisplayText
-                        azureEnvironment.Scopes, accounts.FirstOrDefault()).WithAuthority(msalApp.Authority).ExecuteAsync();
-                }
-                catch (MsalUiRequiredException)
-                {
-                    shouldShowLoginPrompt = true;
-                }
-            }
-
-            if (shouldShowLoginPrompt)
-            {
-                AuthenticationResult = await msalApp.AcquireTokenWithDeviceCode(
-                    azureEnvironment.Scopes,
-                    deviceCodeResult =>
-                    {
-                        channel.Stdout(deviceCodeResult.Message);
-                        return Task.FromResult(0);
-                    }).WithAuthority(msalApp.Authority).ExecuteAsync();
-            }
-
-            if (AuthenticationResult == null)
+            var azureEnvironment = AzureEnvironment.Create(subscriptionId);
+            ActiveWorkspace = await azureEnvironment.GetAuthenticatedWorkspaceAsync(channel, resourceGroupName, workspaceName, refreshCredentials);
+            if (ActiveWorkspace == null)
             {
                 return AzureClientError.AuthenticationFailed.ToExecutionResult();
             }
 
-            var credentials = new Rest.TokenCredentials(AuthenticationResult.AccessToken);
-            QuantumClient = new QuantumClient(credentials)
-            {
-                SubscriptionId = subscriptionId,
-                ResourceGroupName = resourceGroupName,
-                WorkspaceName = workspaceName,
-                BaseUri = azureEnvironment.BaseUri,
-            };
-            ActiveWorkspace = new Azure.Quantum.Workspace(
-                QuantumClient.SubscriptionId,
-                QuantumClient.ResourceGroupName,
-                QuantumClient.WorkspaceName,
-                AuthenticationResult?.AccessToken,
-                azureEnvironment.BaseUri);
-
-            try
-            {
-                AvailableProviders = await QuantumClient.Providers.GetStatusAsync();
-            }
-            catch (Exception e)
+            AvailableProviders = await ActiveWorkspace.GetProvidersAsync();
+            if (AvailableProviders == null)
             {
-                Logger?.LogError(e, $"Failed to download providers list from Azure Quantum workspace: {e.Message}");
                 return AzureClientError.WorkspaceNotFound.ToExecutionResult();
             }
 
-            channel.Stdout($"Connected to Azure Quantum workspace {QuantumClient.WorkspaceName}.");
+            channel.Stdout($"Connected to Azure Quantum workspace {ActiveWorkspace.Name}.");
 
             return ValidExecutionTargets.ToExecutionResult();
         }
 
         /// <inheritdoc/>
         public async Task<ExecutionResult> GetConnectionStatusAsync(IChannel channel)
         {
-            if (QuantumClient == null || AvailableProviders == null)
+            if (ActiveWorkspace == null || AvailableProviders == null)
             {
```

Comment:
I wonder if it's the right message that you can't connect to a workspace unless it has Q# providers.

Reply:
The behavior in this case is that the connection to the workspace will succeed, but there will simply be no valid targets to choose from. But it's a good point, and I'll add an output message to make this clear. It would still be possible, for example, to query the jobs list in this case.